### ISSUE TITLE
`FixedTanimotoGP` class for caching

### ISSUE THREAD

**Issue by wvirany** [AustinT/tanimoto-gp#6]
Hi @AustinT,

To provide an option for caching the Cholesky factor and `K_test_train` matrix, I have been designing a "fixed" GP class, which provides the same functionality as `ConstantMeanTanimotoGP`; however, it assumes that the parameters and test set are fixed. Here's how this would look:

```py
class FixedTanimotoGP:
    """TanimotoGP with fixed mean function. Assumes fixed parameters"""
    def __init__(self, fp_func: Callable[[str], Any], smiles_train: list[str], y_train: jnp.ndarray, amplitude: float, noise: float, mean: float = 0.0):
        self._fp_func = fp_func
        self._amplitude = amplitude
        self._noise = noise
        self._mean = mean
        self._y_centered = self._y_train - mean
        self._K_test_train = None
        self._smiles_test = None
```

Here, the parameters are stored as class attributes, along with the cached data (including the cached Cholesky factor, which is set in the `set_training_data()` method). Other than that, the functionality for the user would be the same.

Let me know if you have any suggestions for how to design this, and if you think this is a reasonable addition to the library (I'll implement it for my own use anyway, since it will be useful for BO). I'll follow up with the full implementation in a PR shortly.

**Comment by AustinT** [comment]
Hi @wvirany, thanks for the suggested class! I think there are actually *two* kinds of fixed GP that could be useful:

- *Fixed training set*: user inputs new *test* points each time, but the parameters and train set are fixed, allowing the Cholesky factor to be cached.
- *Fixed training and test set*: useful for BO over a fixed set of points. Would cache Cholesky factor + train-test covariance matrix

For the purposes of this library I think that the fixed training set might be more useful because it is more general. Why don't you just code whatever you want for your own library and we can think about what might be useful for _this_ library later?

**Comment by wvirany** [comment]
@AustinT Ok that sounds good!
